src/Iterator/RandomIterator.php

```diff
@@ -51,13 +53,25 @@ public function __construct(Iterator $iterator, ?int $seed = null)
 
     public function current()
     {
+        if (!$this->valid()) {
+            throw new BadMethodCallException("Cannot call 'current()' on invalid RandomIterator.");
+        }
+
+        assert(null !== $this->key);
+
         $keyValueTuple = $this->getNextItemAtKey($this->key);
 
         return $keyValueTuple[1];
     }
 
     public function key()
     {
+        if (!$this->valid()) {
+            throw new BadMethodCallException("Cannot call 'key()' on invalid RandomIterator.");
+        }
+
+        assert(null !== $this->key);
```

Comment:
have you considered adding an assertion library to this project, like https://github.com/webmozarts/assert? Together with the [phpstan extension](https://github.com/phpstan/phpstan-webmozart-assert) it can be a powerful tool

Reply:
the assertion is needed here to show code analysers that `$this->key` cannot be null after this point (we've already checked that in `valid()`)

Reply:
No I never considered it.

There is a simple reason to that. I want this library to have no dependency. I want it very simple and out of any external deps.

Reply:
Yep I thought that might've been the reason 😄 . Makes sense, there are ways to get around it 👍